### PR TITLE
support changing the prefix of the path of the url

### PR DESCRIPTION
**Pull request by kujyp**
- Improvement of a known limitation.
The limitation is `It does not support changing the prefix of the path of the url (see #155 for workarounds).`
(https://github.com/pypiserver/pypiserver#known-limitations)
- Ref: https://github.com/pypiserver/pypiserver/issues/155
- Changed the links populated by `request.fullpath`. Now it uses `request.url`.

### Descriptions
There is an issue behind proxy using nginx with the prefix of the url.

In your code some links are populated with a proxy header `X-Forwarded-Host`.
Here's an example. `request.url` is changed by a proxy header `X-Forwarded-Host`.
https://github.com/pypiserver/pypiserver/blob/537034eb4323f0a84e66cb8d73e3b152c9d08744/pypiserver/_app.py#L92-L93

<br/><br/><br/><br/>

But `request.fullpath` isn't changed by a proxy header.
https://github.com/pypiserver/pypiserver/blob/537034eb4323f0a84e66cb8d73e3b152c9d08744/pypiserver/_app.py#L83-L96
fullpath
<br/><br/><br/><br/>

### Tests
I've tested with this nginx.conf file.
```
...
        location /pypi/ {
            proxy_set_header  X-Forwarded-Host $host:$server_port/pypi;
            proxy_set_header  X-Forwarded-Proto $scheme;
            proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header  X-Real-IP $remote_addr;
            proxy_pass        http://0.0.0.0:8001/;
        }
...
```

Full nginx.conf file: https://github.com/kujyp/pypiserver/blob/71ad25ee46f1afdf430d0f038c7a511bd2d3e3f3/proxytest/nginx.conf

I've tested with docker environments.
Here's the test assets for environments. https://github.com/kujyp/pypiserver/commit/71ad25ee46f1afdf430d0f038c7a511bd2d3e3f3
<br/>

I've tested with nginx, but not with other environments.
Let me know, if there are side effects on other environments.

### CODE REVIEW

**Comment by kujyp** [comment]
I added several test cases with `X_FORWARDED_HOST` header, too.

If this PR is against the intent of your code, let me know.

File: pypiserver/_app.py [hunk 1]
```diff
@@ -59,6 +59,13 @@ def log_request():
     log.info(config.log_req_frmt, request.environ)
 
 
+@app.hook('before_request')
+def print_request():
+    parsed = urlparse(request.urlparts.scheme + "://" + request.urlparts.netloc)
+    request.custom_host = parsed.netloc
+    request.custom_fullpath = parsed.path.rstrip('/') + '/' + request.fullpath.lstrip('/')
```
Comment (kujyp) [inline pypiserver/_app.py]: I used the names `request.custom_<property>` for additional properties to explicitly specify that these are not from `bottle.py`.

File: tests/test_core.py [hunk 2]
```diff
@@ -96,7 +96,7 @@ def test_hashfile(tmpdir, algo, digest):
 def test_redirect_prefix_encodes_newlines():
     """Ensure raw newlines are url encoded in the generated redirect."""
     request = Namespace(
-        fullpath='/\nSet-Cookie:malicious=1;'
+        custom_fullpath='/\nSet-Cookie:malicious=1;'
```
Comment (kujyp) [inline tests/test_core.py]: It should be changed, because it uses the `Namespace` instance not the `Bottle` instance.

**Comment by skapin** [comment]
Well, perhapse i'm missing something but i got same behavior in recent build... How to fix it ?

**Pull request merged**